File: src/main/kotlin/org/tendiwa/plane/geometry/straightSkeleton/StraightSkeleton.java
package org.tendiwa.plane.geometry.straightSkeleton;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Multimap;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;
import org.jgrapht.UndirectedGraph;
import org.tendiwa.geometry.points.Point;
import org.tendiwa.geometry.polygons.Polygon;
import org.tendiwa.geometry.segments.Segment;
import org.tendiwa.math.constants.ConstantsKt;
import org.tendiwa.plane.geometry.graphs.constructors.Graph2DConstructorsKt;
import org.tenidwa.collections.utils.Collectors;

public final class StraightSkeleton {
    private static int skeletonNumber = 0;
    final Debug debug = new Debug();
    private final InitialListOfActiveVertices initialLav;
    private final PriorityQueue<SkeletonEvent> queue;
    private final Multimap<Point, Point> arcs = HashMultimap.create();
    private final Polygon polygon;
    private int hash = skeletonNumber++;

    public StraightSkeleton(Polygon polygon) {
        this(polygon, false);
    }

    private StraightSkeleton(Polygon polygon, boolean trustCounterClockwise) {
        this.polygon = EdgePerturbationsKt.getParallelAndPerpendicularEdgesDeflected(
            polygon
        );
//		Utils.printListOfPoints(vertices);

        this.initialLav = new InitialListOfActiveVertices(
            this.polygon.getPoints(),
            trustCounterClockwise
        );
        this.queue = new PriorityQueue<>(initialLav.size());

        // [Obdrzalek 1998, paragraph 2.2, algorithm step 1c]
        initialLav.nodes.forEach(this::queueEventFromNode);
        assert !queue.isEmpty();

        while (!queue.isEmpty()) {
            // Convex 2a
            SkeletonEvent event = queue.poll();
            event.handle(this);
//			debug.drawEventHeight(event);
            assert Boolean.TRUE;
        }
        assert !arcs.isEmpty();
    }

    @Override
    public int hashCode() {
        return hash;
    }

    /**
     * Makes a {@link Node} produce a {@link SkeletonEvent} and adds that event
     * to the event queue if it is not null. It it is null, this method does
     * nothing.
     * @param node A node that produces an event.
     */
    void queueEventFromNode(Node node) {
        SkeletonEvent e = node.computeNearerBisectorsIntersection();
        if (e != null) {
            queue.add(e);
        }
    }

    void outputArc(Point start, Point end) {
        assert start != null;
        assert end != null;
        arcs.put(start, end);
        debug.testForNoIntersection(arcs, start, end);
    }

    public UndirectedGraph<Point, Segment> graph() {
        UndirectedGraph<Point, Segment> graph = Graph2DConstructorsKt.Graph2D();
        for (Map.Entry<Point, Collection<Point>> startToEnds : arcs.asMap().entrySet()) {
            Point start = startToEnds.getKey();
            graph.addVertex(start);
            for (Point end : startToEnds.getValue()) {
                graph.addVertex(end);
                graph.addEdge(start, end);
            }
        }
        return graph;
    }

    public List<Segment> originalEdges() {
        return initialLav.edges;
    }

    public ImmutableSet<Polygon> cap(double depth) {
        if (depth <= -ConstantsKt.getEPSILON()) {
            throw new IllegalArgumentException("Cap depth can't be negative");
        }
        if (depth <= ConstantsKt.getEPSILON()) {
            return ImmutableSet.of(polygon);
        } else {
            return new ShrinkedFront(faces(), depth).polygons();
        }
    }

    public Set<StraightSkeletonFace> faces() {
        return initialLav.nodes.stream()
            .map(node -> node.face().toPolygon())
            .collect(Collectors.toLinkedHashSet());
    }

    void queueEvent(SplitEvent splitEvent) {
        assert splitEvent != null;
        queue.add(splitEvent);
    }
}

File: src/main/kotlin/org/tendiwa/plane/geometry/algorithms/polygons/straightSkeleton/StraightSkeleton.kt
package org.tendiwa.plane.geometry.algorithms.polygons.straightSkeleton

import com.google.common.collect.HashMultimap
import com.google.common.collect.ImmutableSet
import org.jgrapht.UndirectedGraph
import org.tendiwa.math.constants.EPSILON
import org.tendiwa.plane.geometry.graphs.constructors.Graph2D
import org.tendiwa.plane.geometry.points.Point
import org.tendiwa.plane.geometry.polygons.Polygon
import org.tendiwa.plane.geometry.segments.Segment
import java.util.*

class StraightSkeleton private constructor(
    polygon: Polygon,
    trustCounterClockwise: Boolean
) {
    internal val debug = Debug()
    private val initialLav: InitialListOfActiveVertices
    private val queue: PriorityQueue<SkeletonEvent>
    private val arcs = HashMultimap.create<Point, Point>()
    private val polygon: Polygon
    private val hash = skeletonNumber++

    constructor(polygon: Polygon) : this(polygon, false) {
    }

    init {
        this.polygon = polygon.parallelAndPerpendicularEdgesDeflected
        //		Utils.printListOfPoints(vertices);

        this.initialLav = InitialListOfActiveVertices(
            this.polygon,
            trustCounterClockwise
        )
        this.queue = PriorityQueue<SkeletonEvent>(initialLav.size)

        // [Obdrzalek 1998, paragraph 2.2, algorithm step 1c]
        initialLav.nodes.forEach({ this.queueEventFromNode(it) })
        assert(!queue.isEmpty())

        while (!queue.isEmpty()) {
            // Convex 2a
            val event = queue.poll()
            event.handle(this)
            //			debug.drawEventHeight(event);
            assert(java.lang.Boolean.TRUE)
        }
        assert(!arcs.isEmpty)
    }

    override fun hashCode(): Int {
        return hash
    }

    /**
     * Makes a [Node] produce a [SkeletonEvent] and adds that event
     * to the event queue if it is not null. It it is null, this method does
     * nothing.
     * @param node A node that produces an event.
     */
    internal fun queueEventFromNode(node: Node) {
        val e = node.computeNearerBisectorsIntersection()
        if (e != null) {
            queue.add(e)
        }
    }

    internal fun outputArc(start: Point, end: Point) {
        arcs.put(start, end)
        debug.testForNoIntersection(arcs, start, end)
    }

    fun graph(): UndirectedGraph<Point, Segment> {
        val graph = Graph2D()
        for (startToEnds in arcs.asMap().entries) {
            val start = startToEnds.key
            graph.addVertex(start)
            for (end in startToEnds.value) {
                graph.addVertex(end)
                graph.addEdge(start, end)
            }
        }
        return graph
    }

    fun originalEdges(): List<Segment> {
        return initialLav.edges
    }

    fun cap(depth: Double): ImmutableSet<Polygon> {
        if (depth <= -EPSILON) {
            throw IllegalArgumentException("Cap depth can't be negative")
        }
        if (depth <= EPSILON) {
            return ImmutableSet.of(polygon)
        } else {
            return ShrinkedFront(faces(), depth).polygons()
        }
    }

    fun faces(): Set<StraightSkeletonFace> {
        return initialLav.nodes
            .map({ node -> node.face().toPolygon() })
            .toCollection(LinkedHashSet())
    }

    internal fun queueEvent(splitEvent: SplitEvent?) {
        assert(splitEvent != null)
        queue.add(splitEvent)
    }

    companion object {
        private var skeletonNumber = 0
    }
}
